src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Razor.PooledObjects;
+
+/// <summary>
+///  Wraps a pooled <see cref="ImmutableArray{T}.Builder"/> but doesn't allocate it until
+///  it's needed. Note: Dispose this to ensure that the pooled array builder is returned
+///  to the pool.
+/// </summary>
+internal ref struct PooledArrayBuilder<T>
+{
+    private readonly ObjectPool<ImmutableArray<T>.Builder> _pool;
+    private ImmutableArray<T>.Builder? _builder;
+
+    public PooledArrayBuilder()
+        : this(ArrayBuilderPool<T>.Default)
+    {
+    }
+
+    public PooledArrayBuilder(ObjectPool<ImmutableArray<T>.Builder> pool)
+    {
+        _pool = pool;
+    }
+
+    public void Dispose()
+    {
+        ClearAndFree();
+    }
+
+    public int Count
+        => _builder?.Count ?? 0;
+
+    public void Add(T item)
+    {
+        _builder ??= _pool.Get();
+        _builder.Add(item);
+    }
+
+    public void AddRange(IReadOnlyList<T> items)
+    {
+        if (items.Count == 0)
+        {
+            return;
+        }
+
+        _builder ??= _pool.Get();
+        _builder.AddRange(items);
+    }
+
+    public void AddRange(IEnumerable<T> items)
+    {
+        _builder ??= _pool.Get();
+        _builder.AddRange(items);
+    }
+
+    public void ClearAndFree()
+    {
+        if (_builder is { } builder)
+        {
+            _pool.Return(builder);
+            _builder = null;
+        }
+    }
+
+    public readonly ImmutableArray<T> ToImmutable()
+        => _builder?.ToImmutable() ?? ImmutableArray<T>.Empty;
```

Comment:
Could we make this just slightly more complicated to avoid an extra alloc (until we depend on SCI 8.0, then we can make it simpler again by calling `_builder.ToImmutableAndClear()`)? If you take this suggestion, it may be worth renaming to `ToImmutableAndClear()` as well, to be clearer about the semantics, but I'm assuming most of uses of this type will be similar to the usage in the compiler, where we don't realize the same array multiple times in 99.9% of cases.

```suggestion
        {
            if (_builder?.Length is 0 or null) return ImmutableArray<T>.Empty;
            else if (_builder.Length == _builder.Count) return _builder.MoveToImmutable();
            else
            {
                var result = _builder.ToImmutable();
                _builder.Clear();
                return result;
            }
        }
```

Reply:
Agreed! The new .NET 8.0 API is called `DrainToImmutable()`. I'll go ahead and use that name and conditional compile on .NET 8.

Reply:
I keep thinking of the api I proposed, forgetting all the valid arguments about naming 😅